File: scripts/bindings/ggplot_binned_styles.py
"""FIX-GG04: binned count palettes through primary authoring and publication."""
import json,sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes());options=c.export_options(640,360).dpi(96).basis('current')
def descriptor(case):
 kind=case['palette'];control=case['control']
 palette='LineType' if kind=='linetype' else {'Brewer':{'id':'Blues','reverse':False}} if kind=='brewer' else {'Shape':{'solid':kind=='solid'}}
 breaks={'Explicit':[1,3,5,7,9]} if control=='explicit' else {'Explicit':[]} if control in ('empty_breaks','null_breaks') else {'Equal':8 if control=='count_eight' else 16} if control in ('count_eight','count_sixteen') else {'Nice':5}
 limits=[-1,5] if control=='limits' else [None,5] if control=='partial_limits' else None
 return {'function':{'Interpolated':{'normalization':{'Ggplot':{'family':'Linear','domain':[0,1],'reverse':False,'rescaler':'Range'}},'output':'Identity','unknown':{'kind':'Color','value':{'space':'Rgb','channels':{'r':127,'g':127,'b':127,'opacity':1}}} if kind=='brewer' else {'kind':'Missing'}}},'training':'Eligible','ggplot':{'Binned':{'palette':palette,'empty_population':False,'nonfinite_population':False,'limits':limits,'oob':'Squish','breaks':breaks,'right':control!='left'}},'guide':{'Binned':'Automatic'}}
records=[]
for index,case in enumerate(json.loads((ROOT/'fixtures/parity/ggplot2/binned-style-palettes.json').read_text())['cases']):
 owned=[];expected=case['result'];failure=expected.get('error') or expected.get('draw_error')
 try:
  inputs=case['inputs'];kind=case['palette'];channel='LineType' if kind=='linetype' else 'Stroke' if kind=='brewer' else 'Shape'
  data=c.Data.columns({'x':c.column(list(map(float,range(len(inputs)))),kind='float64'),'v':c.column([v or 0 for v in inputs],kind='float64').validity([v is not None for v in inputs])});owned.append(data)
  def layer():return c.points().name('marks') if kind=='brewer' else (c.rule() if kind=='linetype' else c.points()).name('marks').value_scale(channel,'v',descriptor(case))
  mapping=c.aes().x('x').y(1.).x2('x').y2(2.)
  if kind=='brewer':mapping=mapping.stroke('v').stroke_scale('bins').fill('v').fill_scale('bins')
  builder=c.plot(data).profile('Ggplot2_4_0_3').aes(mapping).layer(layer())
  if kind=='brewer':builder=builder.scale(c.color_mapped('bins',descriptor(case)))
  plot=builder.build();owned.append(plot)
  wire=plot.to_json();assert json.loads(wire)['version']==27
  restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
  for state in ('original','edited'):
   current=restored if state=='original' else restored.edit().layer('marks',layer()).build()
   if current is not restored:owned.append(current)
   chart=current.chart();owned.append(chart);actual=chart.semantics()['layers'][0];assert not failure,(index,failure)
   wanted=[v for v in expected['values'] if v is not None or kind=='linetype'];aesthetics=actual.get('aesthetics',[]);styles=actual.get('styles',[])
   assert len(aesthetics)==len(wanted),(index,aesthetics,wanted)
   for row_index,(row,want) in enumerate(zip(aesthetics,wanted)):
    got=row.get(channel)
    if kind in ('solid','hollow'):assert got=={'kind':'Number','value':want},(index,got,want)
    elif kind=='linetype':assert got==({'kind':'Missing'} if want is None else {'kind':'Text','value':want}),(index,got,want)
    else:
     rgb=styles[row_index]['stroke'];want='#7F7F7F' if want=='grey50' else want;assert [rgb[k] for k in ('red','green','blue')]==[int(want[i:i+2],16) for i in (1,3,5)],(index,rgb,want)
   records.append({'index':index,'state':state,'version':27,'styles':actual.get('styles',[]),'aesthetics':aesthetics})
   if case['population']=='many' and case['control']=='count_eight' and state=='original':
    request=output.request(current,options);owned.append(request);frame=request.prepare();owned.append(frame)
    for fmt in ('svg','pdf','png'):(out/f'{kind}.{fmt}').write_bytes(frame.export(fmt))
 except c.ChartError as error:
  diagnostic=json.loads(str(error));assert failure,(index,diagnostic)
  assert diagnostic['code']=='CHART_NUMERICAL_DOMAIN',(index,diagnostic)
  records.append({'index':index,'error':diagnostic['code']})
 finally:
  for value in reversed(owned):value.dispose()
assert len(records)==392,len(records)
(out/'binned-style-records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose();print('PASS Python:',len(records),'binned count palette states; twelve publication files.')

File: crates/chart-core/tests/ggplot_binned_styles.rs
//! FIX-GG04: count palettes reuse trained bin classification and discrete generators.
use chart_core::{
    interpolate::{Number, Value},
    scales::*,
};
fn spec(case: &serde_json::Value) -> chart_core::ChartResult<MappedScaleSpec> {
    let palette = match case["palette"].as_str().unwrap() {
        "solid" => GgplotBinnedPalette::Shape { solid: true },
        "hollow" => GgplotBinnedPalette::Shape { solid: false },
        "linetype" => GgplotBinnedPalette::LineType,
        "brewer" => GgplotBinnedPalette::Brewer {
            id: chromatic::SchemeId::Blues,
            reverse: false,
        },
        _ => unreachable!(),
    };
    let control = case["control"].as_str().unwrap();
    let policy = GgplotBinnedPolicy {
        palette: Some(palette),
        breaks: match control {
            "explicit" => GgplotBreaks::Explicit([1., 3., 5., 7., 9.].map(Number).to_vec()),
            "empty_breaks" | "null_breaks" => GgplotBreaks::Explicit(vec![]),
            "count_eight" => GgplotBreaks::Equal(8.),
            "count_sixteen" => GgplotBreaks::Equal(16.),
            _ => GgplotBreaks::Nice(5.),
        },
        limits: match control {
            "limits" => Some([Some(Number(-1.)), Some(Number(5.))]),
            "partial_limits" => Some([None, Some(Number(5.))]),
            _ => None,
        },
        right: control != "left",
        ..Default::default()
    };
    MappedScaleSpec::authored(ScaleFunctionSpec::Interpolated(InterpolatedScaleSpec {
        normalization: NormalizationSpec::Ggplot {
            timestamp: None,
            family: NumericFamily::Linear,
            domain: [Number(0.), Number(1.)],
            reverse: false,
            rescaler: GgplotRescaler::Range,
        },
        output: ScaleRangeFunction::Identity,
        unknown: if case["palette"] == "brewer" {
            Value::Color(
                chart_core::color::Paint::from_css("#7F7F7F")
                    .unwrap()
                    .value(),
            )
        } else {
            Value::Missing
        },
    }))
    .with_ggplot(GgplotScalePolicy::Binned(Box::new(policy)))?
    .with_guide(GgplotScaleGuide::Binned(GgplotGuideLabels::Automatic))
}
#[test]
fn count_palettes_match_216_reference_builds() {
    let fixture: serde_json::Value = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/binned-style-palettes.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 216);
    for case in cases {
        let inputs: Vec<_> = case["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_f64().map(Number))
            .collect();
        let result = spec(case).and_then(|s| s.trained(&inputs)).and_then(|s| {
            let json = serde_json::to_string(&s).unwrap();
            let scale = MappedScale::new(serde_json::from_str(&json).unwrap())?;
            for input in &inputs {
                scale.numeric(input.map(|v| v.0))?;
            }
            scale.binned_guide_entries(4096, 65536)?;
            Ok(scale)
        });
        let expected = &case["result"];
        if expected.get("error").is_some() {
            assert!(result.is_err(), "{case}");
            continue;
        }
        let scale = result.unwrap_or_else(|e| panic!("{e:?}: {case}"));
        for (input, want) in inputs.iter().zip(expected["values"].as_array().unwrap()) {
            let got = scale.numeric(input.map(|v| v.0)).unwrap();
            let want = if want.is_null() {
                Value::Missing
            } else if let Some(v) = want.as_f64() {
                Value::Number(Number(v))
            } else if case["palette"] == "brewer" {
                Value::Color(
                    chart_core::color::Paint::from_css(if want == "grey50" {
                        "#7F7F7F"
                    } else {
                        want.as_str().unwrap()
                    })
                    .unwrap()
                    .value(),
                )
            } else {
                Value::Text(want.as_str().unwrap().into())
            };
            assert_eq!(got, want, "{case} input {input:?}");
        }
        let guide = scale.binned_guide_entries(4096, 65536);
        if expected["guide"].get("error").is_some() {
            assert!(guide.is_err(), "{case}");
            continue;
        }
        let entries = guide.unwrap().unwrap();
        let cuts = expected["guide"]["breaks"].as_array().unwrap();
        assert_eq!(entries.len(), cuts.len(), "{case}");
        for (got, want) in entries.iter().zip(cuts) {
            assert!(
                (got.transformed.0 - want.as_f64().unwrap()).abs() < 2e-12,
                "{case}"
            );
        }
        assert_eq!(
            entries
                .iter()
                .map(|e| e.label.as_deref())
                .collect::<Vec<_>>(),
            expected["guide"]["labels"]
                .as_array()
                .unwrap()
                .iter()
                .map(|v| v.as_str())
                .collect::<Vec<_>>(),
            "{case}"
        );
    }
}
